File: src/398_Random_pick_index/main.py

import random
from typing import List

class Solution:

    def __init__(self, nums: List[int]):
        self.nums = nums #这里仅仅保存nums，不会引入多余的空间

    def pick(self, target: int) -> int:
        # 就是水塘采样的变种
        cnt, i, rtn = 0, 0, -1
        while i < len(self.nums):
            if self.nums[i] == target:
                # 当满足target条件时再开始采样
                cnt += 1
                # if random.random() % cnt < 1: #因为只取一个，所以等价于==0
                # 等价于在[1,cnt]之间取一个随机值
                rand = int(random.random() * cnt)
                if rand % cnt < 1: #因为只取一个，所以等价于==0
                    rtn = i #保存此时的index
            i += 1 # i循环保证while继续执行
        return rtn

if __name__ == "__main__":
  nums = [1,2,3,3,3]
  gua = Solution(nums)
  for i in range(10):
    rtn = gua.pick(target = 3)
    print("%d = %d" % (i+1, rtn))

File: src/398_Random_pick_index/main.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution1 {
public:
    // 这种思路下的时间复杂度为O(1), 空间复杂度为O(n)
    unordered_map<int, vector<int>> helper;

    Solution1(vector<int> & nums) {
        int len = nums.size();
        for (int i = 0; i < len; ++i) {
            helper[nums[i]].push_back(i);
        }
    }
    
    int pick(int target) {
        auto iter = helper.find(target);
        int val = rand() % ((iter->second).size());
        return iter->second[val];
    }
};

// 网上的解答, 采用的是水塘采用reservoir sampling
// 这里的水塘宽度为1, 因此在操作中省略, 然后结合本题, 遍历时的index要另外计算,
// 即val = target时才认为是index++, 时间复杂度O(n), 空间复杂度O(n)
class Solution {
public:
    vector<int> data;

    Solution(vector<int> & nums) {
        data = nums;
    }
    
    int pick(int target) {
        int cnt = 1, ret = -1;
        for (int i = 0; i < data.size(); ++i) {
            // 当val不等于target时, 都continue, 等价于过滤掉
            if (data[i] != target)
                continue;
            // 一般情况下的水塘采样是与水塘宽度m进行比较. 例如rand()%cnt < m;
            // 但是由于本题水塘宽度m=1, rand%cnt < 1就简化为rand%cnt == 0
            // 然后根据条件记录此时的index, 作为返回值; 
            // 另外注意, cnt是从水塘size以外开始的, 所以本题中cnt初始化为0, 
            // 然后在rand()%cnt之前 ++, 或者先初始化为1, 然后在rand()%cnt之后++
            if (rand()%cnt == 0) {
                ret = i;
            }
            ++cnt;// 切记切记不要忘记更新cnt了, 要不然就不是random了
        }
        return ret;
    }
};

int main() {
    int ret, target;
    vector<int> num1 = {1,2,3,3,3};
    Solution gua(num1);
    ret = gua.pick(target=3);
    cout << "rand value = " << ret << endl;
    ret = gua.pick(target=3);
    cout << "rand value = " << ret << endl;
    ret = gua.pick(target=3);
    cout << "rand value = " << ret << endl;
    return 0;
}

